---
output: github_document
bibliography: ubib.bib
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# updog: Using Parental Data for Offspring Genotyping
[![Linux Build Status](https://travis-ci.org/dcgerard/updog.svg?branch=master)](https://travis-ci.org/dcgerard/updog)
[![Windows Build Status](https://ci.appveyor.com/api/projects/status/github/dcgerard/updog?branch=master&svg=true)](https://ci.appveyor.com/project/dcgerard/updog)
<!-- [![Coverage Status](https://img.shields.io/codecov/c/github/dcgerard/updog/master.svg)](https://codecov.io/github/dcgerard/updog?branch=master) -->
<!-- [![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0) -->

This package will fit an empirical Bayesian procedure to genotype autopolyploid individuals from reduced-representation next-generation sequencing (NGS) data, such as genotyping by sequencing (GBS) [@elshire2011robust] or restriction site-associated DNA sequencing (RAD-Seq) [@baird2008rapid]. For such NGS data there exist other methods for genotyping --- see for example [ebg](https://github.com/pblischak/polyploid-genotyping) [@blischak2017snp] and [TET](http://www.g3journal.org/lookup/suppl/doi:10.1534/g3.117.039008/-/DC1) [@maruki2017genotype]. `updog` adds to this field by:

- Incorporating hierarchical structure into the genotyping procedure by either assuming the population consists of siblings (either an F1 or S1 population), or the population is in Hardy-Weinberg equilibrium. Though see @serang2012efficient and @li2011statistical for other discussion on this.
- Estimating sequencing error for each SNP.
- Accounting for read-mapping bias.
- Incorporating overdispersion, a key indicator for the quality of a SNP.
- Automatically accounting for outlying observations.
- Beautifully plotting the NGS data (see `plot.updog`).
- Running a simple goodness of fit test (see `summary.updog`).
- Simulating under a fitted model (see `rupdog`).

We've included a few SNP's from the data of @shirasawa2017high to show off the features of `updog`. See `snpdata`.

A vignette is available [here](https://dcgerard.github.io/updog/articles/smells_like_updog.html).

Please report any bugs/issues [here](https://github.com/dcgerard/updog/issues).

# Installation

To install, run the following code in R:
``` {r, eval = FALSE}
# install.packages("devtools")
devtools::install_github("dcgerard/updog")
```

# Citation
If you find the methods in this package useful, please cite

> Gerard, D., Ferr&atilde;o L.F.V., Garcia, A.A.F., & Stephens, M. (2018). Harnessing Empirical Bayes and Mendelian Segregation for Genotyping Autopolyploids from Messy Sequencing Data. bioRxiv. doi: [10.1101/281550](https://doi.org/10.1101/281550).

Or, using BibTex:
``` tex
@article {gerard2018harnessing,
	author = {Gerard, David and Ferr{\~a}o, Luis Felipe Ventorim and Garcia, Antonio Augusto Franco and Stephens, Matthew},
	title = {Harnessing Empirical Bayes and Mendelian Segregation for Genotyping Autopolyploids from Messy Sequencing Data},
	year = {2018},
	doi = {10.1101/281550},
	publisher = {Cold Spring Harbor Laboratory},
	URL = {https://www.biorxiv.org/content/early/2018/03/16/281550},
	eprint = {https://www.biorxiv.org/content/early/2018/03/16/281550.full.pdf},
	journal = {bioRxiv}
}
```

# References
