<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using Parental Data for Offspring Genotyping • updog</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="/index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/smells_like_updog.html">Smells Like Updog</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    

    
    
<div class="contents">
<!-- README.md is generated from README.Rmd. Please edit that file -->
<div id="updog-using-parental-data-for-offspring-genotyping" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#updog-using-parental-data-for-offspring-genotyping" class="anchor"></a>updog: Using Parental Data for Offspring Genotyping</h1></div>

<p>This package will fit an empirical Bayesian procedure to genotype autopolyploid individuals using reduced-representation next-generation sequencing (NGS) data, such as genotyping by sequencing (GBS) <span class="citation">[@elshire2011robust]</span> or restriction site-associated DNA sequencing (RAD-Seq) <span class="citation">[@baird2008rapid]</span>. For such NGS data there exist other methods for genotyping — see for example <a href="https://github.com/pblischak/polyploid-genotyping">ebg</a> <span class="citation">[@blischak2017snp]</span> and <a href="http://www.g3journal.org/lookup/suppl/doi:10.1534/g3.117.039008/-/DC1">TET</a> <span class="citation">[@maruki2017genotype]</span>. <code>updog</code> adds to this field by:</p>
<ul>
<li>Incorporating hierarchical structure into the genotyping procedure by either assuming the population consists of siblings (either an F1 or S1 population), or the population is in Hardy-Weinberg equilibrium. Though see <span class="citation">@serang2012efficient</span> and <span class="citation">@li2011statistical</span> for other discussion on this.</li>
<li>Estimating sequencing error for each SNP.</li>
<li>Accounting for read-mapping bias.</li>
<li>Incorporating overdispersion, a key indicator for the quality of a SNP.</li>
<li>Automatically accounting for outlying observations.</li>
<li>Beautifully plotting the NGS data (see <code>plot.updog</code>).</li>
<li>Running a simple goodness of fit test (see <code>summary.updog</code>).</li>
<li>Simulating under a fitted model (see <code>rupdog</code>).</li>
</ul>
<p>We’ve included a few SNP’s from the data of <span class="citation">@shirasawa2017high</span> to show off the features of <code>updog</code>. See <code>snpdata</code>.</p>
<p>Please report any bugs/issues <a href="https://github.com/dcgerard/updog/issues">here</a>.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>To install, run the following code in R:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"dcgerard/updog"</span>)</code></pre></div>
</div>
<div id="citation" class="section level1">
<h1 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h1>
<p>Please cite this package as</p>
<blockquote>
<p>Gerard, David. 2016. <em>updog: Using Parental Data for Offspring Genotyping</em>. <a href="https://github.com/dcgerard/updog" class="uri">https://github.com/dcgerard/updog</a>.</p>
</blockquote>
<p>Or, using BibTex:</p>
<div class="sourceCode"><pre class="sourceCode tex"><code class="sourceCode latex">@Manual{gerard2016updog,
    title = {{updog}: Using Parental Data for Offspring Genotyping},
    author = {David Gerard},
    year = {2016},
    note = {R package version 0.1.0},
    url = {https://github.com/dcgerard/updog},
}</code></pre></div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2 class="hasAnchor">
<a href="#sidebar" class="anchor"></a>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/dcgerard/updog">https://​github.com/​dcgerard/​updog</a>
</li>
<li>Report a bug at <br><a href="http://github.com/dcgerard/updog/issues">http://​github.com/​dcgerard/​updog/​issues</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Gerard <br><small class="roles"> Author, maintainer </small> </li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/dcgerard/updog"><img src="https://travis-ci.org/dcgerard/updog.svg?branch=master" alt="Linux Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/dcgerard/updog"><img src="https://ci.appveyor.com/api/projects/status/github/dcgerard/updog?branch=master&amp;svg=true" alt="Windows Build Status"></a></li>
</ul>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
