<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Smells Like Updog • updog</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/smells_like_updog.html">Smells Like Updog</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Smells Like Updog</h1>
                        <h4 class="author">David Gerard</h4>
            
            <h4 class="date">2018-03-05</h4>
          </div>

    
    
<div class="contents">
<div id="whats-updog" class="section level1">
<h1 class="hasAnchor">
<a href="#whats-updog" class="anchor"></a>What’s Updog?</h1>
<p>Updog is an empirical Bayes approach to genotype individuals (particularly polyploids) from next generation sequencing (NGS) data. We had in mind NGS data that results from a reduced representation library, such as “genotyping-by-sequencing” (GBS) <span class="citation">(Elshire 2011)</span> or “restriction site-associated DNA sequencing” (RAD-seq) <span class="citation">(Baird 2008)</span>.</p>
<p>Updog wields the power of hierarchical modeling to account for some key features of NGS data overlooked in most other analyses: allelic bias, overdispersion, and outlying observations. Updog will also automatically account for sequencing errors.</p>
<p>To efficiently account for these features, updog needs to know the distribution of the individual genotypes in the population. Updog can accurately estimate this distribution if the population consists of full siblings (option <code>model = "s1"</code> or <code>model = "f1"</code>) or if the population is in Hardy-Weinberg equilibrium (<code>model = "hw"</code>).</p>
</div>
<div id="fit-updog" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-updog" class="anchor"></a>Fit Updog</h1>
<p>Load <code>updog</code> and the <code>snpdat</code> dataset. The data frame <code>snpdat</code> contains three example SNPs (single nucleotide polymorphisms) from the study of <span class="citation">Shirasawa et al. (2017)</span>. The individuals in this dataset resulted from a single generation of selfing (an S1 population). You can read more about it by typing <code><a href="../reference/snpdat.html">?snpdat</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)
<span class="kw">library</span>(updog)
<span class="kw">data</span>(<span class="st">"snpdat"</span>)</code></pre></div>
<p>We’ll use the <code>tidyverse</code> to extract the First SNP.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
snpdat <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">filter</span>(snp <span class="op">==</span><span class="st"> "SNP1"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(counts, size, id) -&gt;
<span class="st">  </span>smalldat
<span class="kw">head</span>(smalldat)</code></pre></div>
<pre><code>## # A tibble: 6 x 3
##   counts  size            id
##    &lt;int&gt; &lt;int&gt;         &lt;chr&gt;
## 1    298   354       Xushu18
## 2    187   187 Xushu18S1-001
## 3    201   201 Xushu18S1-002
## 4    157   184 Xushu18S1-003
## 5    175   215 Xushu18S1-004
## 6    283   283 Xushu18S1-005</code></pre>
<p>We will separate the counts between the children and the parent (the first individual). Note that you do <em>not</em> need the parental counts to fit updog, but they can help.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pcounts &lt;-<span class="st"> </span>smalldat<span class="op">$</span>counts[<span class="dv">1</span>]
psize   &lt;-<span class="st"> </span>smalldat<span class="op">$</span>size[<span class="dv">1</span>]
ocounts &lt;-<span class="st"> </span>smalldat<span class="op">$</span>counts[<span class="op">-</span><span class="dv">1</span>]
osize   &lt;-<span class="st"> </span>smalldat<span class="op">$</span>size[<span class="op">-</span><span class="dv">1</span>]
ploidy  &lt;-<span class="st"> </span><span class="dv">6</span> <span class="co"># sweet potatoes are hexaploid</span></code></pre></div>
<p>We can first use <code>plot_geno</code> to visualize the raw data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_geno.html">plot_geno</a></span>(<span class="dt">ocounts =</span> ocounts, <span class="dt">osize =</span> osize, <span class="dt">ploidy =</span> ploidy)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-5-1.png" width="432"></p>
<p>Now we fit <code>updog</code>. We use <code>model = "s1"</code> because the individuals resulted from one generation of selfing of the same parent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">uout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/updog.html">updog</a></span>(<span class="dt">ocounts =</span> ocounts, <span class="dt">osize =</span> osize, <span class="dt">ploidy =</span> ploidy, 
              <span class="dt">p1counts =</span> pcounts, <span class="dt">p1size =</span> psize, <span class="dt">model =</span> <span class="st">"s1"</span>)</code></pre></div>
</div>
<div id="analyzing-output" class="section level1">
<h1 class="hasAnchor">
<a href="#analyzing-output" class="anchor"></a>Analyzing Output</h1>
<p>We use <code>plot.updog</code> to visualize the fit. Points are color coded according to the genotype with the highest posterior probability. For example, a genotype of “4” represents four copies of the reference allele and two copies of the alternative allele (AAAAaa). The level of transparency is proportional to the posterior probability of a point being an outlier. The lines represent the mean counts at a given genotype. The “+” symbol with a black dot is the location of the parent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(uout)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-7-1.png" width="432"></p>
<p>I have incorporated a colorblind safe palate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(uout, <span class="dt">use_colorblind =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-8-1.png" width="432"></p>
<p>The fit seems reasonable. We can test how reasonable our genotype distribution is using <code>summary.updog</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(uout)</code></pre></div>
<pre><code>## $genotypes
## 
##  0  1  2  3  4  5  6 
##  0  0  0  0 39 66 37 
## 
## $summ_prob
##         maxpostprob prob_ok
## Min.         0.5471       1
## 1st Qu.      0.9624       1
## Median       0.9978       1
## Mean         0.9486       1
## 3rd Qu.      0.9999       1
## Max.         1.0000       1
## 
## $gof_pvalue
## [1] 0.6836688</code></pre>
<p>The goodness of fit <span class="math inline">\(p\)</span>-value is quite large, indicating no evidence of lack of fit.</p>
<p>We can sample from the fitted model using <code>rupdog</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rout &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rupdog.html">rupdog</a></span>(uout)</code></pre></div>
<p>If we plot the sampled output, it looks similar to the real data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(rout, <span class="dt">show_outlier =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="smells_like_updog_files/figure-html/unnamed-chunk-11-1.png" width="432"></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-baird2008rapid">
<p>Baird, Paul D. AND Atwood, Nathan A. AND Etter. 2008. “Rapid SNP Discovery and Genetic Mapping Using Sequenced RAD Markers.” <em>PLOS ONE</em> 3 (10). Public Library of Science: 1–7. doi:<a href="https://doi.org/10.1371/journal.pone.0003376">10.1371/journal.pone.0003376</a>.</p>
</div>
<div id="ref-elshire2011robust">
<p>Elshire, Jeffrey C. AND Sun, Robert J. AND Glaubitz. 2011. “A Robust, Simple Genotyping-by-Sequencing (GBS) Approach for High Diversity Species.” <em>PLOS ONE</em> 6 (5). Public Library of Science: 1–10. doi:<a href="https://doi.org/10.1371/journal.pone.0019379">10.1371/journal.pone.0019379</a>.</p>
</div>
<div id="ref-shirasawa2017high">
<p>Shirasawa, Kenta, Masaru Tanaka, Yasuhiro Takahata, Daifu Ma, Qinghe Cao, Qingchang Liu, Hong Zhai, et al. 2017. “A High-Density SNP Genetic Map Consisting of a Complete Set of Homologous Groups in Autohexaploid Sweetpotato (<em>Ipomoea Batatas</em>).” <em>Scientific Reports</em> 7. Nature Publishing Group. doi:<a href="https://doi.org/10.1038/srep44207">10.1038/srep44207</a>.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#whats-updog">What’s Updog?</a></li>
      <li><a href="#fit-updog">Fit Updog</a></li>
      <li><a href="#analyzing-output">Analyzing Output</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
