<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9">
    <div class="page-header">
      <h1>
        Reference
        <small>version&nbsp;0.1.0</small>
      </h1>
    </div>

    <div class="contents">
      <table class="ref-index">

      <colgroup>
        <col class="alias" />
        <col class="title" />
      </colgroup>

      <tbody>
        <tr>
          <th colspan="2">
            <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
            <p class="section-desc"></p>
          </th>
        </tr>
        <tr>
          <!--  -->
          <td>
            <p><code><a href="bbpost_double.html">bbpost_double</a></code> </p>
          </td>
          <td><p>Posterior inference under the beta-binomial model with outliers. If <code>od_param = 0</code>, then
this reduces to the binomial model with outliers.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bb_post.html">bb_post</a></code> </p>
          </td>
          <td><p>Posterior inference in beta-binomial model.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bbpost_tot.html">bbpost_tot</a></code> </p>
          </td>
          <td><p>Posterior inference for each individual.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bb_simple_post.html">bb_simple_post</a></code> </p>
          </td>
          <td><p>Simple posterior inference for beta-binomial when ncounts and ssize are only length 1.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="bin_post.html">bin_post</a></code> </p>
          </td>
          <td><p>Calculates posterior probabilities of a genotype given just the
sequence counts.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="colSums_cpp.html">colSums_cpp</a></code> </p>
          </td>
          <td><p>Stupid implementation of colSums because I guess not implemented in Rcpp sugar.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbetabinom_cpp.html">dbetabinom_cpp</a></code> </p>
          </td>
          <td><p>An Rcpp version of the <code><a href='reference/dbetabinom.html'>dbetabinom</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbetabinom_mu_rho_cpp_double.html">dbetabinom_mu_rho_cpp_double</a></code> </p>
          </td>
          <td><p>double version of betabinomial density.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbetabinom_mu_rho_cpp.html">dbetabinom_mu_rho_cpp</a></code> </p>
          </td>
          <td><p>An Rcpp version of <code><a href='reference/dbetabinom_mu_rho.html'>dbetabinom_mu_rho</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbetabinom_mu_rho.html">dbetabinom_mu_rho</a></code> </p>
          </td>
          <td><p>beta-binomial density function in terms of mean and correlation.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbetabinom.html">dbetabinom</a></code> </p>
          </td>
          <td><p>Beta-binomial density function.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dd.html">dbeta_dd</a></code> </p>
          </td>
          <td><p>Derivative of betabinomial density w.r.t. original bias parameter.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_deps.html">dbeta_deps</a></code> </p>
          </td>
          <td><p>Derivative of beta density w.r.t. sequencing error rate.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dh.html">dbeta_dh</a></code> </p>
          </td>
          <td><p>Derivative of overdispersion parameter of beta density.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dl.html">dbeta_dl</a></code> </p>
          </td>
          <td><p>Derivative of beta density w.r.t. unconstrained parameterization of sequencing error.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dprop.html">dbeta_dprop</a></code> </p>
          </td>
          <td><p>Derivative of mean of beta density.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dr_ell.html">dbeta_dr_ell</a></code> </p>
          </td>
          <td><p>Same as <code><a href='reference/dbeta_dh.html'>dbeta_dh</a></code>, but with same inputs as <code><a href='reference/dbeta_ds.html'>dbeta_ds</a></code>
and <code><a href='reference/dbeta_dl.html'>dbeta_dl</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dr.html">dbeta_dr</a></code> </p>
          </td>
          <td><p>Derivative of beta(x|n, xi, r), where r = log(h) from <code><a href='reference/dbeta_dh.html'>dbeta_dh</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_ds.html">dbeta_ds</a></code> </p>
          </td>
          <td><p>Derivative of betabinomial density w.r.t. bias parameter.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dtau.html">dbeta_dtau</a></code> </p>
          </td>
          <td><p>Derivative of beta(x|n, xi, tau).</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dbeta_dtau_withxi.html">dbeta_dtau_withxi</a></code> </p>
          </td>
          <td><p>Derivative of overdispersion parameter with mean already calculated</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dd_ds.html">dd_ds</a></code> </p>
          </td>
          <td><p>Just a wrapper for <code>std::exp</code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="deps_dell.html">deps_dell</a></code> </p>
          </td>
          <td><p>Derivative of exp(ell) / (1 + exp(ell))</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="df_deps.html">df_deps</a></code> </p>
          </td>
          <td><p>Returns derivative of p(1 - eps) + (1 - p) * eps</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dh_dr.html">dh_dr</a></code> </p>
          </td>
          <td><p>Just a wrapper for <code>std::exp</code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dh_dtau.html">dh_dtau</a></code> </p>
          </td>
          <td><p>Derivative of h(tau) = (1 - tau) / tau.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dhyper_cpp.html">dhyper_cpp</a></code> </p>
          </td>
          <td><p>Vectorized version of <code><a href='http://www.rdocumentation.org/packages/stats/topics/Hypergeometric'>dhyper</a></code> for C++ implementation.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dupdog.html">dupdog</a></code> </p>
          </td>
          <td><p>Returns log-likelihood of <code>updog</code> object.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dxi_dd.html">dxi_dd</a></code> </p>
          </td>
          <td><p>Derivative w.r.t. d of xi(d, f) = f / (d * (1 - f) + f)</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="dxi_df.html">dxi_df</a></code> </p>
          </td>
          <td><p>Returns derivative to f / (d * (1 - f) + f)</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="est_seq_error.html">est_seq_error</a></code> </p>
          </td>
          <td><p>Estimate sequencing error from counts.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="est_seq_post.html">est_seq_post</a></code> </p>
          </td>
          <td><p>Use posterior probabilities to update the estimated sequencing error.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="expit.html">expit</a></code> </p>
          </td>
          <td><p>The expit function.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_dimname.html">get_dimname</a></code> </p>
          </td>
          <td><p>Returns a vector character strings that are all of the possible
combinations of the reference allele and the non-reference allele.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_out_prop.html">get_out_prop</a></code> </p>
          </td>
          <td><p>E step in EM algorithm for a single individual.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_parent_outprop.html">get_parent_outprop</a></code> </p>
          </td>
          <td><p>E-step for the parents.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_prob_geno.html">get_prob_geno</a></code> </p>
          </td>
          <td><p>Rcpp function to get genotype probabilities assuming either
F1 population given parental genotypes, the allele frequencies assuming
Hardy-Weinberg equilibrium, or just a uniform distribution.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_pvec.html">get_pvec</a></code> </p>
          </td>
          <td><p>Gets all possible binomial probabilities for a given ploidy, bias term, and sequencing
error rate.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_q_array_cpp.html">get_q_array_cpp</a></code> </p>
          </td>
          <td><p>Rcpp implementation of <code><a href='reference/get_q_array.html'>get_q_array</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_q_array.html">get_q_array</a></code> </p>
          </td>
          <td><p>Derive the probabilities of an offspring's genotype given its
parental genotypes for all possible combinations of parental and
offspring genotypes.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="get_transition_mat.html">get_transition_mat</a></code> </p>
          </td>
          <td><p>Get the transition probabilities from going from one ploidy to
another assuming a constant error rate.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="good_bb_obj.html">good_bb_obj</a></code> </p>
          </td>
          <td><p>Objective function for non-outlier model</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring_mat_original.html">grad_offspring_mat_original</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring.html'>obj_offspring</a></code> for each individual. This is in the
original parameterization.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring_mat.html">grad_offspring_mat</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring_reparam.html'>obj_offspring_reparam</a></code> for each individual.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring_original.html">grad_offspring_original</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring.html'>obj_offspring</a></code> using the original parameterization.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring.html">grad_offspring</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring_reparam.html'>obj_offspring_reparam</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring_weights_original.html">grad_offspring_weights_original</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring_weights.html'>obj_offspring_weights</a></code> using original parameterization</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_offspring_weights.html">grad_offspring_weights</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_offspring_weights_reparam.html'>obj_offspring_weights_reparam</a></code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_parent.html">grad_parent</a></code> </p>
          </td>
          <td><p>Gradient of <code><a href='reference/obj_parent.html'>obj_parent</a></code> (when <code>outlier = FALSE</code>) with respect to
<code>bias_val</code>, <code>seq_error</code>, and <code>od_param</code>. Basically, this just calculates
derivatives of the log beta-binomial density.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_parent_reparam.html">grad_parent_reparam</a></code> </p>
          </td>
          <td><p>The gradient of <code><a href='reference/obj_parent_reparam.html'>obj_parent_reparam</a></code> with respect to <code>s</code>, <code>ell</code>,
and <code>r</code>, which recall are reparameterizations of the bias, seqeuncing error, and
overdispersion, respectively.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="grad_wrapp_all.html">grad_wrapp_all</a></code> </p>
          </td>
          <td><p>wrapper for <code><a href='reference/grad_offspring_weights.html'>grad_offspring_weights</a></code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="is.updog.html">is.updog</a></code> </p>
          </td>
          <td><p>Tests if its argument is an updog.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="logsumexp.html">logsumexp</a></code> </p>
          </td>
          <td><p>Log-sum-exponential trick that I use all the time.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_offspring.html">obj_offspring</a></code> </p>
          </td>
          <td><p>Objective function for the offspring.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_offspring_reparam.html">obj_offspring_reparam</a></code> </p>
          </td>
          <td><p>Just a reparameterization of <code><a href='reference/obj_offspring.html'>obj_offspring</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_offspring_vec.html">obj_offspring_vec</a></code> </p>
          </td>
          <td><p>Vector of objective functions for offspring.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_offspring_weights.html">obj_offspring_weights</a></code> </p>
          </td>
          <td><p>Same thing as <code><a href='reference/obj_offspring.html'>obj_offspring</a></code>, but each sample's log-density has a weight.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_offspring_weights_reparam.html">obj_offspring_weights_reparam</a></code> </p>
          </td>
          <td><p>Reparameterization of <code><a href='reference/obj_offspring_weights.html'>obj_offspring_weights</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_parent.html">obj_parent</a></code> </p>
          </td>
          <td><p>Parent contribution to objective function.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_parent_reparam.html">obj_parent_reparam</a></code> </p>
          </td>
          <td><p>A reparameterization of <code><a href='reference/obj_parent.html'>obj_parent</a></code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="obj_wrapp_all.html">obj_wrapp_all</a></code> </p>
          </td>
          <td><p>wrapper for <code><a href='reference/obj_offspring_weights_reparam.html'>obj_offspring_weights_reparam</a></code></p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="out_grad_parent.html">out_grad_parent</a></code> </p>
          </td>
          <td><p>Gradient function for <code><a href='reference/dbetabinom_mu_rho_cpp_double.html'>dbetabinom_mu_rho_cpp_double</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="out_grad_wrapp.html">out_grad_wrapp</a></code> </p>
          </td>
          <td><p>Wrapper for <code><a href='reference/outlier_grad.html'>outlier_grad</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="outlier_bb_obj.html">outlier_bb_obj</a></code> </p>
          </td>
          <td><p>Objective function for outlier model.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="outlier_grad.html">outlier_grad</a></code> </p>
          </td>
          <td><p>The gradient of <code><a href='reference/outlier_obj.html'>outlier_obj</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="outlier_obj.html">outlier_obj</a></code> </p>
          </td>
          <td><p>Objective function of outlier part in EM step</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="out_obj_wrapp.html">out_obj_wrapp</a></code> </p>
          </td>
          <td><p>Wrapper for <code><a href='reference/outlier_obj.html'>outlier_obj</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="pbias_double.html">pbias_double</a></code> </p>
          </td>
          <td><p>A double version of <code><a href='reference/pbias.html'>pbias</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="pbias.html">pbias</a></code> </p>
          </td>
          <td><p>Returns the probability of seeing the reference allele after including
the mapping-bias and the sequencing-error.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_beta_dist_gg.html">plot_beta_dist_gg</a></code> </p>
          </td>
          <td><p>Plot the beta distribution using ggplot2.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_beta_dist.html">plot_beta_dist</a></code> </p>
          </td>
          <td><p>Plot the beta distribution.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_geno_base.html">plot_geno_base</a></code> </p>
          </td>
          <td><p>The base R graphics version of <code><a href='reference/plot_geno.html'>plot_geno</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_geno.html">plot_geno</a></code> </p>
          </td>
          <td><p>Make a genotype plot.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot_problines.html">plot_problines</a></code> </p>
          </td>
          <td><p>Plots the genotype lines for a given sequencing error rate, bias parameter, and ploidy.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="plot.updog.html">plot</a></code> </p>
          </td>
          <td><p>Draw a genotype plot from the output of <code><a href='reference/updog.html'>updog</a></code>.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="rbetabinom_mu_rho.html">rbetabinom_mu_rho</a></code> </p>
          </td>
          <td><p>Draw from beta-binomial distribution parameterized
by mean and overdispersion parameter.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="rupdog.html">rupdog</a></code> </p>
          </td>
          <td><p>Samples counts using the parameter values in an <code>updog</code> object.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="screen_oracle.html">screen_oracle</a></code> </p>
          </td>
          <td><p>Returns the distribution of proportion of individuals genotyped correctly using an the oracle estimator
when the parameter values equal the estimated values.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="snpdat.html">snpdat</a></code> </p>
          </td>
          <td><p>GBS data from Shirasawa et al (2017)</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="summary.updog.html">summary</a></code> </p>
          </td>
          <td><p>Summary method for class "<code>updog</code>".</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="up_bb_fix.html">up_bb_fix</a></code> </p>
          </td>
          <td><p>EM iteration for overdispersed binomial mixture model.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="up_bb_obj.html">up_bb_obj</a></code> </p>
          </td>
          <td><p>Objective function for overdispersed binomial mixture model</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="update_good.html">update_good</a></code> </p>
          </td>
          <td><p>Update the OK points</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_iterate.html">updog_iterate</a></code> </p>
          </td>
          <td><p>Iterate between indiividual level estimates and joint averaging.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_maximize.html">updog_maximize</a></code> </p>
          </td>
          <td><p>Estimate parental genotypes using offspring genotypes.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_mcmc.html">updog_mcmc</a></code> </p>
          </td>
          <td><p>The updog Gibbs sampler to jointly estimate the parental and offspring genotypes.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_old.html">updog_old</a></code> </p>
          </td>
          <td><p>Using Parental Data for Offspring Genotyping.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog.html">updog</a></code> </p>
          </td>
          <td><p>Using Parental Data for Offspring Genotyping.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_update_all.html">updog_update_all</a></code> </p>
          </td>
          <td><p>This is the main optimization function for updog.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="updog_vanilla.html">updog_vanilla</a></code> </p>
          </td>
          <td><p>Using Parental Data for Offspring Genotyping.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="up_fix.html">up_fix</a></code> </p>
          </td>
          <td><p>Fixed-point iteration for updog EM.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="up_max_bb.html">up_max_bb</a></code> </p>
          </td>
          <td><p>EM algorithm for over-dispersed mixture model.</p></td>
        </tr><tr>
          <!--  -->
          <td>
            <p><code><a href="up_obj.html">up_obj</a></code> </p>
          </td>
          <td><p>The objective function</p></td>
        </tr>
      </tbody>
      </table>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
