<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Using Parental Data for Offspring Genotyping. — updog • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/smells_like_updog.html">Smells Like Updog</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Using Parental Data for Offspring Genotyping.</h1>
    </div>

    
    <p>This function fits a hierarchical model to sequence counts from
a collection of siblings --- or a population of individuals
in Hardy-Weinberg equilibrium --- and returns genotyped information. The
hierarchy comes from either the fact that they share the same parents or they come
from a population in Hardy-Weinberg equilibrium. If
you also have parental sequencing data, then you can include this
to improve estimates. This is the simple version of <code>updog</code> with only a limited number of parameters.
To see the full list of options, see <code><a href='updog_vanilla.html'>updog_vanilla</a></code>.</p>
    

    <pre class="usage"><span class='fu'>updog</span>(<span class='no'>ocounts</span>, <span class='no'>osize</span>, <span class='no'>ploidy</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"f1"</span>, <span class='st'>"s1"</span>, <span class='st'>"hw"</span>, <span class='st'>"uniform"</span>), <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ocounts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith child.</p></td>
    </tr>
    <tr>
      <th>osize</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads for the ith child.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>A positive integer. The number of copies of the
genome in the species. This is the assumed to be the same
for all individuals.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>The model for the genotype distribution. Do we assume an
F1 population (<code>"f1"</code>), an S1 population (<code>"s1"</code>), Hardy-Weinberg equilibrium (<code>"hw"</code>),
or a uniform distribution (<code>"uniform"</code>).</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments to pass to <code><a href='updog_vanilla.html'>updog_vanilla</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of class <code>updog</code> with some or all of the following elements:</p><dl class='dl-horizontal'>
        <dt><code>ogeno</code></dt><dd><p>A vector. Each element of which is the maximum a posteriori estimate of each individual's genotype.</p></dd>
        <dt><code>maxpostprob</code></dt><dd><p>A vector. The maximum posterior probability of a genotype for each individual.</p></dd>
        <dt><code>postmean</code></dt><dd><p>A vector. The posterior mean genotype for each individual.</p></dd>
        <dt><code>bias_val</code></dt><dd><p>The estimated bias parameter. This is a value greater than 0 which is the ratio of the probability of correctly mapping a read containing the alternative allele to the probability of correctly mapping a read containing the reference allele. A value of 1 indicates no bias. A value less than one indicates bias towards the reference alelle. A value greater than 1 indiciates bias towards the alternative allele.</p></dd>
        <dt><code>seq_error</code></dt><dd><p>The estimated sequencing error rate. This is between 0 and 1.</p></dd>
        <dt><code>od_param</code></dt><dd><p>The estimated overdispersion parameter. Also known as the "intra-class correlation", this is the overdispersion parameter in the underlying beta of the beta-binomial distribution of the counts. Between 0 and 1, a value closer to 0 indicates less overdispersion and a value greater than 1 indicates greater overdispersion. In real data, we we typically see estimates between 0 and 0.01.</p></dd>
        <dt><code>p1geno</code></dt><dd><p>The estimated genotype of one parent. The number of copies of the reference allele one of the parents has.</p></dd>
        <dt><code>p1geno</code></dt><dd><p>The estimated genotype of the other parent. The number of copies of the reference allele the other parent has.</p></dd>
        <dt><code>allele_freq</code></dt><dd><p>The estimated allele-frequency of the reference allele. This is the binomial proportion. Between 0 and 1, a value closer to 1 indicates a larger amount of reference alleles in the population.</p></dd>
        <dt><code>out_prop</code></dt><dd><p>The estimated proportion of points that are outliers.</p></dd>
        <dt><code>out_mean</code></dt><dd><p>The estimated mean of the outlier distribution. The outlier distribution is beta-binomial.</p></dd>
        <dt><code>out_disp</code></dt><dd><p>The estimated overdispersion parameter of the outlier distribution. This is the "intra-class correlation" parameter of the beta-binomial outlier distribution.</p></dd>
        <dt><code>prob_out</code></dt><dd><p>A vector. Each element of which is the posterior probability that a point is an outlier.</p></dd>
        <dt><code>prob_ok</code></dt><dd><p>The posterior probability that a point is a non-outlier.</p></dd>
        <dt><code>p1_prob_out</code></dt><dd><p>The posterior probability that parent 1 is an outlier.</p></dd>
        <dt><code>p2_prob_out</code></dt><dd><p>The posterior probability that parent 2 is an outlier.</p></dd>
        <dt><code>num_iter</code></dt><dd><p>The number of iterations the optimization program was run.</p></dd>
        <dt><code>convergence</code></dt><dd><p><code>1</code> if we reached <code>maxiter</code> and <code>0</code> otherwise.</p></dd>
        <dt><code>llike</code></dt><dd><p>The final log-likelihood of the estimates.</p></dd>
        <dt><code>hessian</code></dt><dd><p>The negative-Fisher information under the parameterization (s, ell, r), where s = log(bias_val) = log(d), ell = logit(seq_error) = logit(eps), and r = - logit(od_param) = - logit(tau). If you want standard errors for these parameters (in the described parameterization), simply take the negative inverse of the hessian.</p></dd>
        <dt><code>input</code></dt><dd><p>A list with the input counts (<code>ocounts</code>), the input sizes (<code>osize</code>), input parental counts (<code>p1counts</code> and <code>p2counts</code>), input parental sizes (<code>p2size</code> and <code>p1size</code>), the ploidy (<code>ploidy</code>) and the model (<code>model</code>).</p></dd>
        <dt><code>log_bias</code></dt><dd><p>The log of <code>bias_val</code></p></dd>
        <dt><code>logit_seq_error</code></dt><dd><p>The logit of <code>seq_error</code>. I.e. <code>log(seq_error / (1 - seq_error))</code></p></dd>
        <dt><code>neg_logit_od_param</code></dt><dd><p>The negative logit of <code>od_param</code>. I.e. <code>log((1 - od_param) / od_param)</code></p></dd>
        <dt><code>covmat</code></dt><dd><p>The observed Fisher information matrix of <code>c(log_bias, logit_seq_error, neg_logit_od_param)</code>. This can be used as the covariance matrix of these estimates.</p></dd>
    </dl>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The key improvements in <code>updog</code> are its abilities to account for common features in
GBS data: sequencing error rate, read-mapping bias, overdispersion, and outlying points.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='updog_vanilla.html'>updog_vanilla</a></code> for more parameter options.
<code><a href='plot.updog.html'>plot.updog</a></code> For plotting the results of <code>updog</code>.
<code><a href='summary.updog.html'>summary.updog</a></code> For some summary capabilities of <code>updog</code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Read in data and format it for SNP1 -------------------------------------</span>
<span class='fu'>data</span>(<span class='no'>snpdat</span>)
<span class='no'>ploidy</span> <span class='kw'>&lt;-</span> <span class='fl'>6</span>
<span class='no'>snp1</span> <span class='kw'>&lt;-</span> <span class='no'>snpdat</span>[<span class='no'>snpdat</span>$<span class='no'>snp</span> <span class='kw'>==</span> <span class='st'>"SNP1"</span>, ]
<span class='no'>ocounts</span> <span class='kw'>&lt;-</span> <span class='no'>snp1</span>$<span class='no'>counts</span>[-<span class='fl'>1</span>]
<span class='no'>osize</span>   <span class='kw'>&lt;-</span> <span class='no'>snp1</span>$<span class='no'>size</span>[-<span class='fl'>1</span>]
<span class='no'>pcounts</span> <span class='kw'>&lt;-</span> <span class='no'>snp1</span>$<span class='no'>counts</span>[<span class='fl'>1</span>]
<span class='no'>psize</span>   <span class='kw'>&lt;-</span> <span class='no'>snp1</span>$<span class='no'>size</span>[<span class='fl'>1</span>]

<span class='co'>## Fit updog ---------------------------------------------------------------</span>
<span class='co'>## Here, I don't update the parental genotype, but you should use the</span>
<span class='co'>## default setting, `update_pgeno = TRUE`, in practice.</span>
<span class='no'>uout</span> <span class='kw'>&lt;-</span> <span class='fu'>updog</span>(<span class='kw'>ocounts</span> <span class='kw'>=</span> <span class='no'>ocounts</span>, <span class='kw'>osize</span> <span class='kw'>=</span> <span class='no'>osize</span>, <span class='kw'>p1counts</span> <span class='kw'>=</span> <span class='no'>pcounts</span>,
              <span class='kw'>p1size</span> <span class='kw'>=</span> <span class='no'>psize</span>, <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>6</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='st'>"s1"</span>, <span class='kw'>p1geno</span> <span class='kw'>=</span> <span class='fl'>5</span>,
              <span class='kw'>p2geno</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>update_pgeno</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
<span class='fu'>plot</span>(<span class='no'>uout</span>)</div><div class='img'><img src='updog-1.png' alt='' width='700' height='432.632880098887' /></div><div class='input'><span class='fu'>summary</span>(<span class='no'>uout</span>)</div><div class='output co'>#&gt; $genotypes
#&gt; 
#&gt;  0  1  2  3  4  5  6 
#&gt;  0  0  0  0 39 66 37 
#&gt; 
#&gt; $summ_prob
#&gt;         maxpostprob prob_ok
#&gt; Min.          0.547       1
#&gt; 1st Qu.       0.962       1
#&gt; Median        0.998       1
#&gt; Mean          0.949       1
#&gt; 3rd Qu.       1.000       1
#&gt; Max.          1.000       1
#&gt; 
#&gt; $gof_pvalue
#&gt; [1] 0.6837
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
