<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>This is the main optimization function for updog. — updog_update_all • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>This is the main optimization function for updog.</h1>
    </div>

    
    <p>This is the main optimization function for updog.</p>
    

    <pre class="usage"><span class='fu'>updog_update_all</span>(<span class='no'>ocounts</span>, <span class='no'>osize</span>, <span class='no'>ploidy</span>, <span class='kw'>p1counts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p1size</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>p2counts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p2size</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>print_val</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>tol</span> <span class='kw'>=</span> <span class='fl'>10</span>^-<span class='fl'>4</span>,
  <span class='kw'>maxiter</span> <span class='kw'>=</span> <span class='fl'>500</span>, <span class='kw'>commit_num</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>min_disp</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>update_outmean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>update_outdisp</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>update_outprop</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_bias_val</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>update_seq_error</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_od_param</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_pgeno</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>p1geno</span> <span class='kw'>=</span> <span class='fu'>round</span>(<span class='no'>ploidy</span>/<span class='fl'>2</span>), <span class='kw'>p2geno</span> <span class='kw'>=</span> <span class='fu'>round</span>(<span class='no'>ploidy</span>/<span class='fl'>2</span>), <span class='kw'>seq_error</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
  <span class='kw'>od_param</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>bias_val</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>out_prop</span> <span class='kw'>=</span> <span class='fl'>0.001</span>, <span class='kw'>out_mean</span> <span class='kw'>=</span> <span class='fl'>0.5</span>,
  <span class='kw'>out_disp</span> <span class='kw'>=</span> <span class='fl'>1</span>/<span class='fl'>3</span>, <span class='kw'>non_mono_max</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>bound_bias</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>seq_error_mean</span> <span class='kw'>=</span> -<span class='fl'>4.7</span>, <span class='kw'>seq_error_sd</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>bias_val_mean</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>bias_val_sd</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>allele_freq</span> <span class='kw'>=</span> <span class='fl'>0.5</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"f1"</span>, <span class='st'>"s1"</span>, <span class='st'>"hw"</span>,
  <span class='st'>"uniform"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ocounts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith child.</p></td>
    </tr>
    <tr>
      <th>osize</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads for the ith child.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>A positive integer. The number of copies of the
genome in the species. This is the assumed to be the same
for all individuals.</p></td>
    </tr>
    <tr>
      <th>p1counts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith
sample of parent 1.  If <code>NULL</code> then the prior
probabilities on parent 1's genotype will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p1size</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads in the ith sample of parent 1.  If
<code>NULL</code> then the prior probabilities on parent 1's genotype
will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p2counts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith
sample of parent 2.  If <code>NULL</code> then the prior
probabilities on parent 2's genotype will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p2size</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads in the ith sample of parent 2.  If
<code>NULL</code> then the prior probabilities on parent 2's genotype
will default to uniform.</p></td>
    </tr>
    <tr>
      <th>print_val</th>
      <td><p>A logical. Should we print the updates (<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>tol</th>
      <td><p>The stopping criterion</p></td>
    </tr>
    <tr>
      <th>maxiter</th>
      <td><p>The maximum number of iterations</p></td>
    </tr>
    <tr>
      <th>commit_num</th>
      <td><p>The number of consecutive iterations where the parental
genotypes do not change before we commit to those parental genotypes.</p></td>
    </tr>
    <tr>
      <th>min_disp</th>
      <td><p>The minimum value for the over-dispersion parameter of the
outlier distribution. If this gets too small, then we can be overly confident in
points being outliers.</p></td>
    </tr>
    <tr>
      <th>update_outmean</th>
      <td><p>A logical. Should we update <code>out_mean</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_outdisp</th>
      <td><p>A logical. Should we update <code>out_mean</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_outprop</th>
      <td><p>A logical. Should we update <code>out_prop</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_bias_val</th>
      <td><p>A logical. Should we update <code>bias_val</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_seq_error</th>
      <td><p>A logical. Should we update <code>seq_error</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_od_param</th>
      <td><p>A logical. Should we update <code>od_param</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>update_pgeno</th>
      <td><p>A logical. Should we update <code>p1geno</code> and <code>p1geno</code>
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>p1geno</th>
      <td><p>The initial value of the first parental genotype.</p></td>
    </tr>
    <tr>
      <th>p2geno</th>
      <td><p>The initial value of the second parental genotype.</p></td>
    </tr>
    <tr>
      <th>seq_error</th>
      <td><p>The initial value of the sequencing error rate.</p></td>
    </tr>
    <tr>
      <th>od_param</th>
      <td><p>The initial value of the overdispersion parameter.</p></td>
    </tr>
    <tr>
      <th>bias_val</th>
      <td><p>The initial value of the bias parameter.</p></td>
    </tr>
    <tr>
      <th>out_prop</th>
      <td><p>The initial value of the proportion of points that are outliers.</p></td>
    </tr>
    <tr>
      <th>out_mean</th>
      <td><p>The initial value of the mean of the outlier distribution.</p></td>
    </tr>
    <tr>
      <th>out_disp</th>
      <td><p>The initial value of the over-dispersion parameter of the outlier distribution.</p></td>
    </tr>
    <tr>
      <th>non_mono_max</th>
      <td><p>The maximum number of iterations to allow non-monotonicity of likelihood.</p></td>
    </tr>
    <tr>
      <th>bound_bias</th>
      <td><p>A logical. Should we bound <code>bias_val</code> by a somewhat arbitrary value
(<code>TRUE</code>) or not (<code>FALSE</code>)?</p></td>
    </tr>
    <tr>
      <th>seq_error_mean</th>
      <td><p>The mean of the logit-normal prior on the sequencing error rate, which corresponds
to <code>parvec[2]</code>. Set <code>seq_error_sd = Inf</code> to have no penalty on the sequencing error rate.
The default is -4.7, which roughly corresponds to a mean sequencing error value of 0.009.</p></td>
    </tr>
    <tr>
      <th>seq_error_sd</th>
      <td><p>The standard deviation of the logit-normal prior on the sequencing error rate, which
corresponds to <code>parvec[2]</code>. The default is 1, which at three standard deviations is about a sequencing
error rate of 0.15.</p></td>
    </tr>
    <tr>
      <th>bias_val_mean</th>
      <td><p>The prior mean on the log of <code>bias_val</code> (corresponding to <code>parvec[1]</code>).
Set <code>bias_val_sd = Inf</code> to have no penalty on the bias parameter.</p></td>
    </tr>
    <tr>
      <th>bias_val_sd</th>
      <td><p>The prior standard deviation on the log of <code>bias_val</code>
(corresponding to <code>parvec[1]</code>).
Set <code>bias_val_sd = Inf</code> to have no penalty on the bias parameter.</p></td>
    </tr>
    <tr>
      <th>allele_freq</th>
      <td><p>The allele frequency if <code>model = "hw"</code>.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>The model for the genotype distribution. Do we assume an
F1 population (<code>"f1"</code>), an S1 population (<code>"s1"</code>), Hardy-Weinberg equilibrium (<code>"hw"</code>),
or a uniform distribution (<code>"uniform"</code>).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list of the following elements</p><ul>
<li><p><code>bias_val</code>: The estimated bias parameter.</p></li>
<li><p><code>seq_error</code>: The estimated sequencing error rate.</p></li>
<li><p><code>od_param</code>: The estimated overdispersion parameter.</p></li>
<li><p><code>p1geno</code>: The estimated genotype of one parent.</p></li>
<li><p><code>p1geno</code>: The estimated genotype of the other parent.</p></li>
<li><p><code>out_prop</code>: The estimated proportion of points that are outliers.</p></li>
<li><p><code>out_mean</code>: The estimated mean of the outlier distribution.</p></li>
<li><p><code>out_disp</code>: The estimated overdispersion parameter of the outlier distribution.</p></li>
<li><p><code>prob_out</code>: A vector. Each element of which is the posterior probability that a point is an outlier.</p></li>
<li><p><code>allele_freq</code>: The estimated allele-frequency of the reference allele.</p></li>
<li><p><code>p1_prob_out</code>: The posterior probability that parent 1 is an outlier.</p></li>
<li><p><code>p2_prob_out</code>: The posterior probability that parent 2 is an outlier.</p></li>
<li><p><code>num_iter</code>: The number of iterations the optimization program was run.</p></li>
<li><p><code>convergence</code>: 1 is we reached <code>maxiter</code> and 0 otherwise.</p></li>
<li><p><code>llike</code>: The final log-likelihood of the estimates.</p></li>
<li><p><code>hessian</code>: The Fisher information under the parameterization (s, ell, r), where s = log(bias_val) = log(d),
ell = logit(seq_error) = logit(eps), and r = - logit(od_param) =  - logit(tau).</p></li>
</ul>

    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
