<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Using Parental Data for Offspring Genotyping. — updog_old • updog</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">updog</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dcgerard/updog">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Using Parental Data for Offspring Genotyping.</h1>
    </div>

    
    <p>This function fits a hierarchical model to sequence counts from a
collection of siblings and returns genotyped information. The
hierarchy comes from the fact that they share the same parents. If
you also have parental sequencing data, then you can include this
to improve estimates.</p>
    

    <pre class="usage"><span class='fu'>updog_old</span>(<span class='no'>ocounts</span>, <span class='no'>osize</span>, <span class='no'>ploidy</span>, <span class='kw'>p1counts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p1size</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>p2counts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>p2size</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>seq_error</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>integrate</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>do_eb</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>overdispersion</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_geno</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>update_pi</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_outlier</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>update_rho</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ocounts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith child.</p></td>
    </tr>
    <tr>
      <th>osize</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads for the ith child.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>A positive integer. The number of copies of the
genome in the species. This is the assumed to be the same
for all individuals.</p></td>
    </tr>
    <tr>
      <th>p1counts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith
sample of parent 1.  If <code>NULL</code> then the prior
probabilities on parent 1's genotype will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p1size</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads in the ith sample of parent 1.  If
<code>NULL</code> then the prior probabilities on parent 1's genotype
will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p2counts</th>
      <td><p>A vector of non-negative integers. The ith element
is the number of reads of the reference allele in the ith
sample of parent 2.  If <code>NULL</code> then the prior
probabilities on parent 2's genotype will default to uniform.</p></td>
    </tr>
    <tr>
      <th>p2size</th>
      <td><p>A vector of positive integers. The ith element is the
total number of reads in the ith sample of parent 2.  If
<code>NULL</code> then the prior probabilities on parent 2's genotype
will default to uniform.</p></td>
    </tr>
    <tr>
      <th>seq_error</th>
      <td><p>A non-negative numeric. This is the known
sequencing error rate. The default is to estimate this
using data that are all approximately the reference allele.</p></td>
    </tr>
    <tr>
      <th>integrate</th>
      <td><p>A logical. Should we integrate over our
uncertainty in the parental genotypes (<code>TRUE</code>) or not
(<code>FALSE</code>). The default is <code>FALSE</code> because we usually
know the parental genotypes with near certainty so it's not
important to integrate over our uncertainty in them. This is only
implemented if <code>do_eb = FALSE</code></p></td>
    </tr>
    <tr>
      <th>do_eb</th>
      <td><p>Should we do empirical Bayes (<code>TRUE</code>) or not (<code>FALSE</code>)?
You should have a lot of parental data to be able to set this to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>overdispersion</th>
      <td><p>A logical. Should we fit a beta-binomial model
to account for overdispersion (<code>TRUE</code>) or not (<code>FALSE</code>)? If
<code>overdispersion = TRUE</code> then we start the overdispersion parameter,
<code>rho</code> at 0.001, a very small value. If parental information is
provided, then we use that data as the starting values for <code>rho</code>. If
<code>update_rho</code> is <code>FALSE</code>, then these values are fixed throughout the
estimation procedure.</p></td>
    </tr>
    <tr>
      <th>update_geno</th>
      <td><p>A logical. Update the parental genotypes? If <code>FALSE</code> and if you have parental data, then
we fix the parental genotypes to be the maximum a posteriori values. If you do not have parental data,
then this should not be set to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>update_pi</th>
      <td><p>A logical. Update the mixing proporiton? If <code>FALSE</code>, then
1% of the observations are assumed to be outliers.</p></td>
    </tr>
    <tr>
      <th>update_outlier</th>
      <td><p>A logical. Update the outlier distribution? I <code>FALSE</code>, then
the outlier distribution is assumed to just be a uniform from 0 to 1.</p></td>
    </tr>
    <tr>
      <th>update_rho</th>
      <td><p>A logical. Update the overdispersion parameter?</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with some or all of the following elements:</p>
<p><code>opostprob</code>: A matrix of proportions whose (i, j)th
    element is the posterior probability that child j has i - 1
    copies of the reference allele.  That is, the rows index the
    genotype and the columns index the offspring.</p>
<p><code>p1postprob</code>: A vector of proportions whose ith element is
  the posterior probability that parent 1 has i - 1 copies of the
  reference allele.  These are derived ONLY from parent 1's
  sequence data and not jointly with all of the data.</p>
<p><code>p2postprob</code>: A vector of proportions whose ith element is
  the posterior probability that parent 2 has i - 1 copies of the
  reference allele.  These are derived ONLY from parent 2's
  sequence data and not jointly with all of the data.</p>
<p><code>pival</code>: The estimated proportion of observations that are not outliers.</p>
<p><code>rho</code>: The overdispersion parameter.</p>
<p><code>out_mu</code>: The outlier distribution mean.</p>
<p><code>out_rho</code>: The outlier distribution overdispersion parameter.</p>
<p><code>p1geno</code>: The estimated genopype of parent 1.</p>
<p><code>p2geno</code>: The estimated genotype of parent 2.</p>
<p><code>prob_ok</code>: A vector of proportions. The ith element is the posterior
      probability that the ith element is not an outlier.</p>
<p><code>ogeno</code>: A vector of integers. The ith element is the estimated genotype
      of the ith offspring.</p>
<p><code>alpha</code>: The outlier distributions's shape 1 parameter.</p>
<p><code>beta</code>: The outlier distributions's shape 2 parameter.</p>
<p><code>seq_error</code>: The sequencing error rate used during the updog iterates.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If you have a lot of parental sequencing data, then it could
suffice to run <code>updog</code> with <code>update_geno</code> set to <code>FALSE</code>, which
would save a lot of time.
Otherwise, you will probably want to borrow strength
between the offspring by setting <code>update_geno</code> to <code>TRUE</code>.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
          </ul>

    <h2>Author</h2>
    
David Gerard

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
