# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Objective function for the offspring.
#'
#' @param ocounts A vector of integers. The observed counts of the refernce
#'     allele for each individual.
#' @param osize A vector of integers. The observed number of reads for each
#'     individuals.
#' @param ploidy An integer. The ploidy of the species. This is assumed
#'     to be the same for each individual.
#' @param p1geno The first parental genotype. The number of copies of the
#'     reference allele the first parent has.
#' @param p2geno The second parental genotype. The number of copies of the
#'     reference allele the second parent has.
#' @param bias_val The bias parameter. A value of 1 means there is no bias
#'     toward one allele or the other. A value less than one indicates a bias
#'     toward the reference allele. A value greater than one indicates a bias
#'     toward the non-reference allele.
#' @param seq_error The sequencing error rate.
#' @param od_param The overdispersion parameter in the beta-binomial model
#'     for the OK counts. When this is zero, this resorts to the binomial
#'     model for counts.
#' @param outlier A logical. Should we include an outlier model (\code{TRUE})
#'     or not (\code{FALSE})? Defaults to \code{FALSE}.
#' @param out_prop The proportion of points that are outliers. Defaults
#'     (quite arbitrarily) to \code{0.01}.
#' @param out_mean The mean of beta-binomial for the outlier distribution.
#'     Defaults to \code{0.5}.
#' @param out_disp The overdispersion parameter of the outlier distribution.
#'     Defaults to \code{1/3}, which corresponds to a uniform distribution
#'     for the underlying beta when \code{out_mean = 0.5}.
#'
obj_offspring <- function(ocounts, osize, ploidy, p1geno, p2geno, bias_val = 1, seq_error = 0, od_param = 0, outlier = FALSE, out_prop = 0.01, out_mean = 0.5, out_disp = 1.0 / 3.0) {
    .Call('updog_obj_offspring', PACKAGE = 'updog', ocounts, osize, ploidy, p1geno, p2geno, bias_val, seq_error, od_param, outlier, out_prop, out_mean, out_disp)
}

#' Returns the probability of seeing the reference allele after including
#' the mapping-bias and the sequencing-error.
#'
#' @param prob A numeric vector. Each element is the proportion of genomes that contain
#'     the reference allele. This should take on
#'     one of the values 0/K, 1/K, ... , K/K, where K is the ploidy of the individual.
#' @param bias The bias parameter. Should be greater than or equal to zero,
#'     though is typically less than 1. A 1 indicates no bias. A value less than one
#'     indicates a bias towards the reference allele. A value greater than 1 indicates
#'     a bias towards the non-reference allele.
#' @param seq_error The sequencing error rate. This should be between 0 and 1.
#'
#' @author David Gerard
#'
pbias <- function(prob, bias, seq_error) {
    .Call('updog_pbias', PACKAGE = 'updog', prob, bias, seq_error)
}

#' An Rcpp version of the \code{\link{dbetabinom}}.
#'
#' @param x A numeric vector of the observed counts.
#' @param size A numeric vector of the number of trials.
#' @param alpha_shape A numeric scalar of the alpha parameter underlying beta.
#' @param beta_shape A numeric scalar of the beta parameter of the underlying beta.
#' @param return_log A logical scalar. Should we return the log of the density (\code{TRUE})
#'     or not (\code{FALSE})?
#'
#' @export
#'
#' @author David Gerard
#'
dbetabinom_cpp <- function(x, size, alpha_shape, beta_shape, return_log = FALSE) {
    .Call('updog_dbetabinom_cpp', PACKAGE = 'updog', x, size, alpha_shape, beta_shape, return_log)
}

#' An Rcpp version of \code{\link{dbetabinom_mu_rho}}.
#'
#' @inheritParams dbetabinom_cpp
#' @param mu The mean of the underlying beta.
#' @param rho The overdispersion parameter of the underlying beta.
#'
#' @author David Gerard
#'
dbetabinom_mu_rho_cpp <- function(x, size, mu, rho, return_log = FALSE) {
    .Call('updog_dbetabinom_mu_rho_cpp', PACKAGE = 'updog', x, size, mu, rho, return_log)
}

#' Vectorized version of \code{\link[stats]{dhyper}} for C++ implementation.
#'
#' @param x Vector of quantiles representing the number of white balls drawn without replacement from an urn which contains both black and white balls.
#' @param m The number of white balls in the urn.
#' @param n The number of black balls in the urn.
#' @param k The number of balls drawn from the urn.
#'
#' @author David Gerard
#'
dhyper_cpp <- function(x, m, n, k) {
    .Call('updog_dhyper_cpp', PACKAGE = 'updog', x, m, n, k)
}

#' Rcpp implementation of \code{\link{get_q_array}}.
#'
#' This function will return the segregation proabilities
#'
#' @inheritParams get_q_array
#'
#' @seealso \code{\link{get_q_array}}.
#'
#' @author David Gerard
#'
#' @export
#'
get_q_array_cpp <- function(ploidy) {
    .Call('updog_get_q_array_cpp', PACKAGE = 'updog', ploidy)
}

#' Log-sum-exponential trick that I use all the time.
#'
#' @param xx A matrix whose rows are to be log-sum-exponentiated.
#'
#' @author David Gerard
#'
logsumexp <- function(xx) {
    .Call('updog_logsumexp', PACKAGE = 'updog', xx)
}

